### PR TITLE
Implement better logic for `ApplyStateChange` in skin editor

### PR DESCRIPTION
**Pull request by Terochi**
This PR improves on the naive approach of `ISerialisableDrawableContainer.Reload()` when traversing through history of changes to keep as many components as possible rather than creating new ones.

Also fixes the first state when opening editor/changing layer not being saved.

### CODE REVIEW

File: osu.Game/Overlays/SkinEditor/SkinEditor.cs [hunk 1]
```diff
@@ -479,12 +477,18 @@ private void revert()
 
         protected void Cut()
         {
+            if (!canCut.Value)
+                return;
```
Comment (bdach) [inline osu.Game/Overlays/SkinEditor/SkinEditor.cs:481]: What are these changes doing here?

File: osu.Game.Tests/Visual/Gameplay/TestSceneSkinEditor.cs [hunk 2]
```diff
@@ -97,15 +100,10 @@ public void TestDragSelection()
                 InputManager.PressButton(MouseButton.Left);
             });
 
-            AddStep("Drag to bottom right", () =>
-            {
-                InputManager.MoveMouseTo(box3.ScreenSpaceDrawQuad.TopRight + new Vector2(-box3.ScreenSpaceDrawQuad.Width / 8, box3.ScreenSpaceDrawQuad.Height / 4));
-            });
+            AddStep("Drag to bottom right",
+                () => { InputManager.MoveMouseTo(box3.ScreenSpaceDrawQuad.TopRight + new Vector2(-box3.ScreenSpaceDrawQuad.Width / 8, box3.ScreenSpaceDrawQuad.Height / 4)); });
```
Comment (bdach) [inline osu.Game.Tests/Visual/Gameplay/TestSceneSkinEditor.cs:104]: Please refrain from making these types of reformats of otherwise unrelated code in PRs. They obscure the diff. I have to go check that this isn't a change in logic.

File: osu.Game/Overlays/SkinEditor/SkinEditor.cs [hunk 3]
```diff
@@ -345,6 +342,7 @@ private void targetChanged(ValueChangedEvent<SkinComponentsContainerLookup?> tar
             changeHandler = new SkinEditorChangeHandler(skinComponentsContainer);
             changeHandler.CanUndo.BindValueChanged(v => undoMenuItem.Action.Disabled = !v.NewValue, true);
             changeHandler.CanRedo.BindValueChanged(v => redoMenuItem.Action.Disabled = !v.NewValue, true);
+            changeHandler.SaveState();
```
Comment (bdach) [inline osu.Game/Overlays/SkinEditor/SkinEditor.cs:345]: The better way to do this would likely entail doing this in `SkinEditorChangeHandler` itself. Likely requires changing

https://github.com/ppy/osu/blob/6a04708a7e9801949c6c7ac7ddf6a4d7fa0835e5/osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs#L36

to `BindValueChanged(..., true)`. This way you won't have to duplicate this same call in the test as well.

File: osu.Game.Tests/Visual/Gameplay/TestSceneSkinEditor.cs [hunk 4]
```diff
@@ -182,6 +171,64 @@ public void TestCyclicSelection()
...
+            AddStep("Bring to front", () => skinEditor.BringSelectionToFront());
+            revertAndCheckUnchanged();
+
+            AddStep("Remove components", () => testComponents.ForEach(c => firstTarget.Remove(c, false)));
+            revertAndCheckUnchanged();
+
+            void revertAndCheckUnchanged()
+            {
+                AddStep("Revert changes", () => changeHandler.RestoreState(int.MinValue));
+                AddAssert("Nothing changed", () => defaultState.SequenceEqual(changeHandler.GetCurrentState()));
```
Comment (bdach) [inline osu.Game.Tests/Visual/Gameplay/TestSceneSkinEditor.cs:228]:
```suggestion
                AddAssert("Current state is same as default", () => defaultState.SequenceEqual(changeHandler.GetCurrentState()));
```

File: osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs [hunk 5]
```diff
@@ -56,20 +57,51 @@ protected override void ApplyStateChange(byte[] previousState, byte[] newState)
...
+                int typeComponentsUsed = typedComponentCounter[lookup]++;
+
+                ISerialisableDrawable component;
+
+                if (typeComponentsUsed < typeComponents.Count)
+                    // Re-use unused component
+                    ((Drawable)(component = typeComponents[typeComponentsUsed])).ApplySerialisedInfo(skinnableInfo);
+                else
+                    // Create new one
+                    component = (ISerialisableDrawable)skinnableInfo.CreateInstance();
```
Comment (bdach) [inline osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs:102]: This doesn't need to be this complicated (the stuff with the two dictionaries and the counting). You have a mutable lookup (realised as a `Dictionary<Type, List<ISerialisableDrawable>>`, so use it instead rather than counting manually:

```csharp
if (typeComponents[lookup].Any())
{
    component = typeComponents.Remove(typeComponents.Count - 1);
    ((Drawable)component).ApplySerialisedInfo(skinnableInfo);
}
else
{
    component = (ISerialisableDrawable)skinnableInfo.CreateInstance();
}
```

Hell if you made it a `Dictionary<Type, Stack<ISerialisableDrawable>>` then you wouldn't need to faff about with the indices either, you would just `.Pop()` an instance for reuse.

File: osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs [hunk 6]
```diff
@@ -56,20 +57,51 @@ protected override void ApplyStateChange(byte[] previousState, byte[] newState)
...
 
-                foreach (var drawable in targetComponents)
-                    drawable.ApplySerialisedInfo(skinnableInfo[i++]);
+                firstTarget.Add(component);
```
Comment (bdach) [inline osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs:104]: The method can't end here. The drawables that were not reused need to be disposed, because you removed them with `(..., false)`.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/SkinEditor/SkinEditor.cs (hunk 1 above)
Comment (Terochi) [reply osu.Game/Overlays/SkinEditor/SkinEditor.cs:481]: They are a bit unrelated, but not wrong?

File: osu.Game.Tests/Visual/Gameplay/TestSceneSkinEditor.cs (hunk 2 above)
Comment (Terochi) [reply osu.Game.Tests/Visual/Gameplay/TestSceneSkinEditor.cs:104]: Sorry, this happens whenever I refactor code with my IDE, I'll try to avoid this happening in the future.

File: osu.Game/Overlays/SkinEditor/SkinEditor.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Overlays/SkinEditor/SkinEditor.cs:481]: They may not be wrong but the changes are out of scope of this pull. The relevant menu items already check the `can{Cut,Copy,Paste}` bindables to determine if the menu items should be disabled. Move these changes to a separate pull, with test coverage, if the lack of these checks in this concrete place causes an actual issue.

File: osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs [hunk 7]
```diff
@@ -56,20 +57,56 @@ protected override void ApplyStateChange(byte[] previousState, byte[] newState)
...
-            SerialisedDrawableInfo[] skinnableInfo = deserializedContent.ToArray();
-            Drawable[] targetComponents = firstTarget.Components.OfType<Drawable>().ToArray();
+            SerialisedDrawableInfo[] skinnableInfos = deserializedContent.ToArray();
+            ISerialisableDrawable[] targetComponents = firstTarget.Components.ToArray();
 
-            if (!skinnableInfo.Select(s => s.Type).SequenceEqual(targetComponents.Select(d => d.GetType())))
+            // Store components based on type for later lookup
+            var typedComponents = new Dictionary<Type, Stack<Drawable>>();
+
+            for (int i = targetComponents.Length - 1; i >= 0; i--)
```
Comment (bdach) [inline osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs:66]: Is this indexing backwards because of the stack inverting order? If so, then that I didn't foresee, but you could fix this by using `Queue<T>` which is FIFO rather than `Stack<T>` which is LIFO.

File: osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs [hunk 8]
```diff
@@ -33,7 +34,7 @@ public SkinEditorChangeHandler(Drawable targetScreen)
                 return;
 
             components = new BindableList<ISerialisableDrawable> { BindTarget = firstTarget.Components };
-            components.BindCollectionChanged((_, _) => SaveState());
+            components.BindCollectionChanged((_, _) => SaveState(), true);
```
Comment (Terochi) [inline osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs:37]: Just discovered this cannot be here, as it saves blank state when loading components for the first time.

Comment (bdach) [reply osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs:37]: Okay, well, then I guess this part has to be reverted.

Comment (Terochi) [reply osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs:37]: Seems that putting the binding into `LoadComplete()` fixes it. I don't have enough knowledge about the framework to know if `SkinEditorChangeHandler.LoadComplete()` gets always called after `SkinBlueprintContainer.LoadComplete()`. In that case it should be fine.

Comment (bdach) [reply osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs:37]: We usually put `BindValueChanged()` calls in `LoadComplete()` yes, but in this particular case it's likely that doing so would be nothing less than a crude timing hack. There's really no reason why the proper components shouldn't be present in the target screen at the point of ctor - and if there are issues with components not being present at that time, it should be the screens' responsibility to signal when it is valid to save the initial state rather than have the change handler attempt to guess based on its own lifetime.

Comment (Terochi) [reply osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs]: > Seems that putting the binding into LoadComplete() fixes it.

Don't worry about this, turns out I had some other changed that I was testing with, to not confuse you as in why it should work. Still gonna make separate PR.

**Review by bdach — APPROVED** [review]
I've applied a few stylistic cleanups, but looks good otherwise.
